src/ccs/ir/circuit.rs

```diff
@@ -0,0 +1,579 @@
+            public_inputs: Default::default(),
+        }
+    }
+
+    pub fn write(
+        &mut self,
+        inputs: &HashMap<(usize, UUID), F>,
+        witnesses: &[(UUID, HashMap<UUID, F>)],
+        assign_pos: &AssignmentOffsets,
+    ) {
+        assert_eq!(inputs.len(), self.l);
+        assert_eq!(assign_pos.len(), self.n);
+        let witness_len = self.n - self.l - 1;
+        let mut witness_values = vec![F::ZERO; witness_len];
+        let mut public_values = vec![F::ZERO; self.l];
+        for ((step_idx, signal_id), idx) in assign_pos.iter() {
+            if *signal_id == 0 {
+                continue;
+            }
+            if *idx < self.n - self.l {
+                witness_values[*idx - 1] = *witnesses[*step_idx].1.get(signal_id).unwrap();
+            } else {
+                public_values[*idx - witness_len - 1] =
+                    *inputs.get(&(*step_idx, *signal_id)).unwrap();
+            }
+        }
+        self.assignments = witness_values;
+        self.public_inputs = public_values;
+    }
+
+    pub fn write_with_values(&mut self, inputs: &[F], witnesses: &[F]) {
+        assert_eq!(inputs.len(), self.l);
+        assert_eq!(witnesses.len(), self.n - self.l - 1);
+
+        self.public_inputs.clone_from(&inputs.to_owned());
+        self.assignments.clone_from(&witnesses.to_owned());
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Matrix<F> {
+    n: usize,
+    m: usize,
+    values: Vec<Vec<F>>,
+}
+
+impl<F: Field> Matrix<F> {
+    pub fn new(n: usize, m: usize) -> Self {
+        Self {
+            n,
+            m,
+            values: vec![vec![F::ZERO; n]; m],
+        }
+    }
+
+    pub fn write(&mut self, values: &[(usize, usize, F)]) {
+        for &(row, col, value) in values.iter() {
+            assert!(row < self.m);
+            assert!(col < self.n);
+            self.values[row][col] = value;
+        }
+    }
+
+    pub fn get(&self, row: usize, col: usize) -> F {
+        assert!(row < self.m);
+        assert!(col < self.n);
+        self.values[row][col]
+    }
+
+    pub fn reduce_rows(&mut self, m: usize) {
+        if m < self.m {
+            // self.values = self.values[0..m].to_owned();
+            self.values.truncate(m);
+            self.m = m;
+        }
+    }
+
+    pub fn size(&self) -> (usize, usize) {
+        (self.m, self.n)
+    }
+
+    pub fn values(&self) -> Vec<Vec<F>> {
+        self.values.clone()
+    }
+}
+
+impl<F: Field + From<u64> + Hash> Matrix<F> {
+    pub fn matrix_vector_product(&self, z: &[F]) -> Vec<F> {
+        (0..self.values.len())
+            .map(|i| self.hadamard_product_and_sum(i, z))
+            .collect()
+    }
+
+    fn hadamard_product_and_sum(&self, index: usize, z: &[F]) -> F {
+        assert!(index < self.values.len());
+        assert_eq!(self.values[index].len(), z.len());
+        hadamard_product(&self.values[index], z).iter().sum()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CCSCircuit<F> {
```

Comment:
If this is an IR should have not so much logic, it is just a container for informaation, the compiler should create this information in the unit and then set it in this IR.